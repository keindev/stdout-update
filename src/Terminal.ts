export class Terminal {
  static readonly COLUMNS = 80;
  static readonly EOL = '\n';
  static readonly ROWS = 24;

  readonly #isWin32 = process.platform === 'win32';
  readonly #stdout: NodeJS.WriteStream;

  constructor(stdout: NodeJS.WriteStream) {
    this.#stdout = stdout;
  }

  get width(): number {
    return this.#stdout.columns ? this.adapt(this.#stdout.columns) : Terminal.COLUMNS;
  }

  get height(): number {
    return this.#stdout.rows ? this.adapt(this.#stdout.rows) : Terminal.ROWS;
  }

  adapt(value: number): number {
    return this.#isWin32 ? value - 1 : value;
  }
}
