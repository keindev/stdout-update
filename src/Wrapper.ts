import ansiStyles from 'ansi-styles';
import stringWidth from 'string-width';
import stripAnsi from 'strip-ansi';

import { Terminal } from './Terminal.js';

const ESCAPES = new Set(['\u001B', '\u009B']);
const DEFAULT_COLOR_CODE = 39;
const INDENT = 4;

export class Wrapper {
  #rows: string[];

  constructor() {
    this.#rows = [''];
  }

  wrap(str: string, limit: number): string[] {
    if (!str.trim().length) return [''];

    this.#rows = [''];

    const rows = this.#rows;
    let rowLength: number;
    let wordLength: number;

    str
      .normalize()
      .split(' ')
      .forEach((word, index): void => {
        rowLength = stringWidth(rows[rows.length - 1] ?? '');
        wordLength = stringWidth(word);

        if (index !== 0) {
          rows[rows.length - 1] += ' ';
          rowLength++;
        }

        if (wordLength > limit) {
          const remainingColumns = limit - rowLength;
          const breaksStartingThisLine = 1 + Math.floor((wordLength - remainingColumns - 1) / limit);
          const breaksStartingNextLine = Math.floor((wordLength - 1) / limit);

          if (breaksStartingNextLine < breaksStartingThisLine) rows.push('');

          this.wrapWord(word, limit);
        } else {
          if (rowLength && wordLength && rowLength + wordLength > limit) rows.push('');

          rows[rows.length - 1] += word;
        }
      });

    return this.wrapEscapes(rows.map(value => value.trimRight()).join(Terminal.EOL)).split(Terminal.EOL);
  }

  private wrapEscapes(characters: string): string {
    const slice = (index: number): RegExpExecArray | null => /\d[^m]*/.exec(characters.slice(index, index + INDENT));
    const wrap = (code: number): string => `${ESCAPES.values().next().value}[${code}m`;
    let result = '';
    let match: RegExpExecArray | null;
    let code: number | undefined;
    let escapeCode: number | undefined;

    [...characters].forEach((character, index): void => {
      result += character;

      if (ESCAPES.has(character)) {
        match = slice(index);

        if (match && match[0]) {
          code = parseFloat(match[0]);
          escapeCode = code === DEFAULT_COLOR_CODE ? undefined : code;
        }
      }

      if (escapeCode) {
        code = ansiStyles.codes.get(escapeCode);

        if (code) {
          if (characters[index + 1] === Terminal.EOL) result += wrap(code);
          if (character === Terminal.EOL) result += wrap(escapeCode);
        }
      }
    });

    return result;
  }

  private wrapWord(word: string, limit: number): void {
    const rows = this.#rows;
    let isInsideEscape = false;
    let visible = stringWidth(stripAnsi(rows[rows.length - 1] ?? ''));

    [...word].forEach((character, index, characters): void => {
      const characterLength = stringWidth(character);

      if (visible + characterLength <= limit) {
        rows[rows.length - 1] += character;
      } else {
        rows.push(character);
        visible = 0;
      }

      if (ESCAPES.has(character)) isInsideEscape = true;
      else if (isInsideEscape && character === 'm') isInsideEscape = false;
      else if (!isInsideEscape) {
        visible += characterLength;

        if (visible === limit && index < characters.length - 1) {
          rows.push('');
          visible = 0;
        }
      }
    });

    if (!visible && (rows[rows.length - 1] ?? '').length > 0 && rows.length > 1) rows[rows.length - 2] += rows.pop();
  }
}
