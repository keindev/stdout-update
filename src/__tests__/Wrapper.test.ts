import stripAnsi from 'strip-ansi';

import { Terminal } from '../Terminal.js';
import { Wrapper } from '../Wrapper.js';

let wrapper: Wrapper;
const red = (s: string): string => `\u001B[31m${s}\u001b[39m`;
const wrap = (s: string, l: number): string => wrapper.wrap(s, l).join(Terminal.EOL);
const mapWrap = (a: string[], l: number): string => a.map((s): string => wrap(s, l)).join(Terminal.EOL);
const FIXTURE = `The quick brown ${red('fox jumped over ')}the lazy ${red('dog and then ran away with the quokka.')}`;

describe('Wrapper', (): void => {
  beforeEach((): void => {
    wrapper = new Wrapper();
  });

  it('Breaks strings longer than "cols" characters', (): void => {
    const long = wrapper.wrap(FIXTURE, 20);

    expect(long).toMatchSnapshot();
    expect(long.every((line): boolean => stripAnsi(line).length <= 20)).toBeTruthy();

    const short = wrapper.wrap(FIXTURE, 5);

    expect(short).toMatchSnapshot();
    expect(short.every((line): boolean => stripAnsi(line).length <= 5)).toBeTruthy();
  });

  it('Word-wrapping and no trimming', (): void => {
    expect(mapWrap(['12345678', '901234567890 12345'], 13)).toBe('12345678\n901234567890\n12345');
    expect(mapWrap(['12345678', '901234567890'], 10)).toBe('12345678\n9012345678\n90');
    expect(wrap('12345678', 5)).toBe('12345\n678');
    expect(wrap('12345678 ', 5)).toBe('12345\n678');
    expect(wrap('   ', 2)).toBe('');
    expect(wrap('foo bar', 3)).toBe('foo\nbar');
    expect(wrap('안녕하세', 4)).toBe('안녕\n하세');
  });

  it('Removes last row if it contained only ansi escape codes', (): void => {
    expect(
      stripAnsi(wrap(red('absolution'), 2))
        .split('\n')
        .every((line): boolean => line.length === 2)
    ).toBeTruthy();
  });

  it('Does not prepend newline if first word is split', (): void => {
    const result = wrapper.wrap(`${red('hello')}world`, 5);

    expect(result.length).toBe(2);
  });

  it('Supports unicode surrogate pairs', (): void => {
    expect(wrap('a\uD83C\uDE00bc', 2)).toBe('a\n\uD83C\uDE00\nbc');
    expect(wrap('a\uD83C\uDE00bc\uD83C\uDE00d\uD83C\uDE00', 2)).toBe(
      'a\n\uD83C\uDE00\nbc\n\uD83C\uDE00\nd\n\uD83C\uDE00'
    );
  });
});
