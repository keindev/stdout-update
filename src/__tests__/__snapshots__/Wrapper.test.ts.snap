// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper Breaks strings longer than "cols" characters 1`] = `
[
  "The quick brown [31mfox[39m",
  "[31mjumped over [39mthe lazy",
  "[31mdog and then ran[39m",
  "[31maway with the[39m",
  "[31mquokka.[39m",
]
`;

exports[`Wrapper Breaks strings longer than "cols" characters 2`] = `
[
  "The",
  "quick",
  "brown",
  "[31mfox j[39m",
  "[31mumped[39m",
  "[31mover[39m",
  "[31m[39mthe",
  "lazy",
  "[31mdog[39m",
  "[31mand[39m",
  "[31mthen[39m",
  "[31mran[39m",
  "[31maway[39m",
  "[31mwith[39m",
  "[31mthe[39m",
  "[31mquokk[39m",
  "[31ma.[39m",
]
`;
